gpii/node_modules/userListeners/test/userFolderTests.js

```diff
@@ -0,0 +1,290 @@
+                    currentTest.expect, "reject");
+
+                doTest(index + 1);
+            });
+        }
+    };
+
+    doTest(0);
+});
+
+jqUnit.asyncTest("userFolder writeTokenFile", function () {
+
+    var tempDir = os.tmpdir();
+    var tokenFile = "test-token-file" + Math.random();
+    var tokenPath = path.join(tempDir, tokenFile);
+
+    teardowns.push(function () {
+        try {
+            fs.unlinkSync(tokenPath);
+        } catch (e) {
+            // ignore
+        }
+    });
+
+    fluid.promise.sequence([
+        function () {
+            var p = gpii.userListeners.writeTokenFile(tempDir, tokenFile, "token1", true);
+            jqUnit.assertTrue("1st writeTokenFile should return a promise", fluid.isPromise(p));
+
+            return p.then(function () {
+                var content = fs.readFileSync(tokenPath, "utf8");
+                jqUnit.assertEquals("Token in file should match what was written", "token1", content);
+            });
+        },
+        function () {
+            // Using the same path as before also tests over-writing.
+            var p = gpii.userListeners.writeTokenFile(tokenPath, null, "token2", true);
+            jqUnit.assertTrue("2nd writeTokenFile should return a promise", fluid.isPromise(p));
+
+            return p.then(function () {
+                var content = fs.readFileSync(tokenPath, "utf8");
+                jqUnit.assertEquals("Token in file should match what was written", "token2", content);
+            });
+        },
+        function () {
+            var p = gpii.userListeners.writeTokenFile("does/not/exist", tokenFile, "token3", true);
+            jqUnit.assertTrue("3rd writeTokenFile should return a promise", fluid.isPromise(p));
+
+            var promiseTogo = fluid.promise();
+            p.then(function () {
+                jqUnit.fail("writeTokenFile with a bad path should reject");
+                promiseTogo.reject();
+            }, promiseTogo.resolve);
+        }
+    ]).then(jqUnit.start, jqUnit.fail);
+});
+
+// Test if the login on startup gets called.
+jqUnit.asyncTest("Test attemptLogin", function () {
+
+    jqUnit.expect(3);
+
+    var testToken = "test-token" + Math.random();
+
+    // Write the test token file.
+    var tempDir = os.tmpdir();
+    var tokenFile = "test-token-file" + Math.random();
+    var tokenPath = path.join(tempDir, tokenFile);
+    fs.writeFileSync(tokenPath, testToken, "utf8");
+
+    teardowns.push(function () {
+        try {
+            fs.unlinkSync(tokenPath);
+        } catch (e) {
+            // ignore
+        }
+    });
+
+    var userFolder = gpii.tests.userListener.getUserFolderListener();
+
+    userFolder.tokenDirectory = tempDir;
+    userFolder.tokenFile = tokenFile;
+
+    var eventRaised = false;
+
+    userFolder.events.onTokenArrive.addListener(function (that, token) {
+        jqUnit.assertFalse("Token event should only fire once", eventRaised);
```

Comment:
Take the chance to test the status of `that.done` too.

It would be nice to have a test case for when `that.getToken()` cannot read a GPII key.